resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala

```diff
@@ -502,6 +521,25 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
     )
   }
 
+  private def satisfiesLocality(offerHostname: String): Boolean = {
+    if (hostToLocalTaskCount.nonEmpty) {
```

Comment:
Now that I'm thinking more about this, it's a bit tricker than I originally thought.

The problem with the current PR is that, while it works fine if all tasks are submitted at once, it won't have any benefit for tasks submitted after the scheduler starts, which is what happens in multi-stage jobs.  Since you're measuring the delay timeout from the time the scheduling starts, then we'll accept all offers that come after that delay, regardless of locality.  So if the timeout is 3s, and the scheduler receives a request to schedule a task 5s after the job starts, then it will launch an executor to run that task on whatever offer happens to arrive first.

What we *really* want is, for each task submitted to the `TaskScheduler`, we want to start counting the delay timeout from the moment the task was submitted.  However, the `ExecutorAllocationClient` interface doesn't provide us with task IDs.  It only gives us hostnames via `hostsToLocalTaskCount`.

I think we can approximate the behavior we want by storing a per-host timestamp representing the last time its entry in `hostsToLocalTaskCount` was modified.  Then we only accept an offer for a different host if enough time has elapsed past this stored time. 

There's code in `YarnAllocator.scala` that does similar things that might theoretically provide inspiration, but I just looked and found it to be quite unreadable.

We should also only do this if dynamic allocation is enabled, so we should probably just condition the entire check on `Utils.isDynamicAllocationEnabled`.